Support creation of draft releases and draft PRs

gpetiot opened the pull request:
~#262 should be merged before~ done

Will fix #246 

- [x] Add a `--draft` option to `dune-release publish` so that the release is created as a draft.
- [x] Add a `--draft` option to `dune-release opam submit` so that the PR is opened as a draft.
- [x] Add a `dune-release undraft` command:
  - [x] writing the release id in a file upon creation so that it can be retrieved while undrafting the release
  - [x] undraft the release on github
  - [x] update the opam file with the new release url
  - [x] writing the pr number in a file upon creation so that it can be retrieved while undrafting the PR
  - [x] undraft the opam-repo PR on github
  - [x] check that we don't try to publish a draft release in a non-draft PR

gpetiot commented on lib/curl.ml:
```diff
@@ -2,47 +2,9 @@ open Bos_setup
   let json =
-    strf "{ \"tag_name\" : \"%s\", \"body\" : \"%s\" }"
-      (escape_for_json version) (escape_for_json msg)
+    strf {|{ "tag_name" : %S, "body" : %S, "draft" : %b }|} version msg draft
```
It seems the escape function is not  needed anymore when using `{| ... |}`

NathanReb commented:
Hmm it seems the changelog check isn't working as intended :/

NathanReb reviewed (commented):
I give this a quick spin and I have a few notes:
- When running in draft mode, I think it would be nice if the logs were adjusted e.g. mentioned creating a draft release, opening a draft PR etc...
- The temp files used to store the PR and Release ids are written at the root of the project as `.draft_pr` and `.draft_release`. This triggers a dune-release warning about the repo being dirty and it is not ideal. I think they should instead be written to the `_build` directory and probably in the same way than the `.url` file, i.e. as `_build/<package_name>-<version>.draft_pr` so that those files can be specific to a certain release at least. We'll probably have to improve the robustness with a centralized tmp folder for dune-release but I think that'd be a good first step.
- The undraft didn't work, I haven't taken the time to debug it yet but if you want to take a look, here's the error message:
```
$ dune-release undraft
[-] Undrafting release
dune-release: [ERROR] Could not find "message" from:
  `Assoc ([("url",
            `String ("https://api.github.com/repos/NathanReb/dune-release-testing/releases/34826727"));
            ("assets_url",
             `String ("https://api.github.com/repos/NathanReb/dune-release-testing/releases/34826727/assets"));
            ("upload_url",
             `String ("https://uploads.github.com/repos/NathanReb/dune-release-testing/releases/34826727/assets{?name,label}"));
            ("html_url",
             `String ("https://github.com/NathanReb/dune-release-testing/releases/tag/v0.2.0%2Btest"));
            ("id", `Int (34826727));
            ("author",
             `Assoc ([("login", `String ("NathanReb")); ("id", `Int (7419360));
                       ("node_id", `String ("MDQ6VXNlcjc0MTkzNjA="));
                       ("avatar_url",
                        `String ("https://avatars2.githubusercontent.com/u/7419360?v=4"));
                       ("gravatar_id", `String (""));
                       ("url",
                        `String ("https://api.github.com/users/NathanReb"));
                       ("html_url", `String ("https://github.com/NathanReb"));
                       ("followers_url",
                        `String ("https://api.github.com/users/NathanReb/followers"));
                       ("following_url",
                        `String ("https://api.github.com/users/NathanReb/following{/other_user}"));
                       ("gists_url",
                        `String ("https://api.github.com/users/NathanReb/gists{/gist_id}"));
                       ("starred_url",
                        `String ("https://api.github.com/users/NathanReb/starred{/owner}{/repo}"));
                       ("subscriptions_url",
                        `String ("https://api.github.com/users/NathanReb/subscriptions"));
                       ("organizations_url",
                        `String ("https://api.github.com/users/NathanReb/orgs"));
                       ("repos_url",
                        `String ("https://api.github.com/users/NathanReb/repos"));
                       ("events_url",
                        `String ("https://api.github.com/users/NathanReb/events{/privacy}"));
                       ("received_events_url",
                        `String ("https://api.github.com/users/NathanReb/received_events"));
                       ("type", `String ("User"));
                       ("site_admin", `Bool (false))]));
            ("node_id", `String ("MDc6UmVsZWFzZTM0ODI2NzI3"));
            ("tag_name", `String ("v0.2.0+test"));
            ("target_commitish", `String ("master")); ("name", `Null);
            ("draft", `Bool (false)); ("prerelease", `Bool (false));
            ("created_at", `String ("2020-12-04T13:44:28Z"));
            ("published_at", `String ("2020-12-04T13:57:38Z"));
            ("assets",
             `List ([`Assoc ([("url",
                               `String ("https://api.github.com/repos/NathanReb/dune-release-testing/releases/assets/29148260"));
                               ("id", `Int (29148260));
                               ("node_id",
                                `String ("MDEyOlJlbGVhc2VBc3NldDI5MTQ4MjYw"));
                               ("name", `String ("dummy-v0.2.0.test.tbz"));
                               ("label", `String (""));
                               ("uploader",
                                `Assoc ([("login", `String ("NathanReb"));
                                          ("id", `Int (7419360));
                                          ("node_id",
                                           `String ("MDQ6VXNlcjc0MTkzNjA="));
                                          ("avatar_url",
                                           `String ("https://avatars2.githubusercontent.com/u/7419360?v=4"));
                                          ("gravatar_id", `String (""));
                                          ("url",
                                           `String ("https://api.github.com/users/NathanReb"));
                                          ("html_url",
                                           `String ("https://github.com/NathanReb"));
                                          ("followers_url",
                                           `String ("https://api.github.com/users/NathanReb/followers"));
                                          ("following_url",
                                           `String ("https://api.github.com/users/NathanReb/following{/other_user}"));
                                          ("gists_url",
                                           `String ("https://api.github.com/users/NathanReb/gists{/gist_id}"));
                                          ("starred_url",
                                           `String ("https://api.github.com/users/NathanReb/starred{/owner}{/repo}"));
                                          ("subscriptions_url",
                                           `String ("https://api.github.com/users/NathanReb/subscriptions"));
                                          ("organizations_url",
                                           `String ("https://api.github.com/users/NathanReb/orgs"));
                                          ("repos_url",
                                           `String ("https://api.github.com/users/NathanReb/repos"));
                                          ("events_url",
                                           `String ("https://api.github.com/users/NathanReb/events{/privacy}"));
                                          ("received_events_url",
                                           `String ("https://api.github.com/users/NathanReb/received_events"));
                                          ("type", `String ("User"));
                                          ("site_admin", `Bool (false))]));
                               ("content_type", `String ("application/x-tar"));
                               ("state", `String ("uploaded"));
                               ("size", `Int (1117));
                               ("download_count", `Int (0));
                               ("created_at", `String ("2020-12-04T13:46:17Z"));
                               ("updated_at", `String ("2020-12-04T13:46:17Z"));
                               ("browser_download_url",
                                `String ("https://github.com/NathanReb/dune-release-testing/releases/download/v0.2.0%2Btest/dummy-v0.2.0.test.tbz"))
                               ])
                      ]));
            ("tarball_url",
             `String ("https://api.github.com/repos/NathanReb/dune-release-testing/tarball/v0.2.0+test"));
            ("zipball_url",
             `String ("https://api.github.com/repos/NathanReb/dune-release-testing/zipball/v0.2.0+test"));
            ("body", `String ("CHANGES:\n\n- Extract library\n"))])
```

gpetiot commented:
So apparently when updating the release online the `browser_download_url` field is not returned, we would only get one after uploading a release, so we would need to remove the old one to reupload the new one but we can't guess the url to put it in the opam file before we resubmit it, or maybe I missed something?

I moved the `.draft_pr`/`.draft_release` to the build directory by the way.

NathanReb commented:
> So apparently when updating the release online the `browser_download_url` field is not returned, we would only get one after uploading a release, so we would need to remove the old one to reupload the new one but we can't guess the url to put it in the opam file before we resubmit it, or maybe I missed something

The new url is indeed in the response, see this section from the above error:
```
("browser_download_url",
                                `String ("https://github.com/NathanReb/dune-release-testing/releases/download/v0.2.0%2Btest/dummy-v0.2.0.test.tbz"))
```
It appears to be under `.assets.[0].browser_download_url` in the json response. We need a way to detect which asset is the right one but we can use the file name of the asset and we could eventually set the label when uploading it in `publish` so we can use it here to retrieve it!

gpetiot commented:
I've fixed this issue when undrafting the release (still having a look at the issues rising when undrafting the PR)

gpetiot commented:
@NathanReb I implemented the requests to undraft pull requests using the github v4 graphQL API

The following workflow now works without any issue: (replace with your own fork of opam-repo)
```sh
$ dune-release distrib --yes
$ dune-release publish --draft --yes
$ dune-release opam pkg
$ dune-release opam submit --opam-repo gpetiot/opam-repository --draft --yes
$ dune-release undraft --opam-repo gpetiot/opam-repository --yes
```

gpetiot commented:
I've rebased the branch, made sure we use the sanitized tag for the branch name, and added a hint to use --draft when opening a pr for a draft release.

gpetiot commented:
I've rebased it again, let me know if it would be better to split this PR into smaller bits, and especially the v3/v4 layer and all the tests.

gpetiot commented:
@NathanReb @pitag-ha let me know when we can merge this one!

NathanReb commented on tests/bin/draft/dune:
```diff
@@ -0,0 +1,3 @@
+(mdx
```
Could use cram tests for new tests?

gpetiot replied:
If I remember correctly if we enable the new cram feature we have to upgrade all existing tests as we can't say for which .t file it should be enabled or disabled.

NathanReb replied:
Ah ok, make sense, let's stick to mdx then!

NathanReb commented on tests/bin/draft/run.t:
```diff
@@ -0,0 +1,128 @@
+(1) distrib
+
+    $ dune-release distrib --dry-run
+    [-] Building source archive
```
The output of this command can be ignored, we just want to create the tarball for the next command right?

gpetiot replied:
Indeed, done!

NathanReb commented on bin/cli.ml:
```diff
@@ -148,6 +148,13 @@ let dry_run =
 
+let draft =
+  let doc =
+    "Produce a draft release that cannot be merged accidentally and has to be \
```
How about we flag this as experimental so people don't expect the behaviour to be too stable for now?

gpetiot replied:
done!

NathanReb reviewed (commented):
It looks great! Thanks!

I have a few comments regarding the `undraft` implementation. It is mostly one big function and it's pretty hard to follow what's in there. Could you try to extract named functions to clarify what are the actions we're taking?
I also noted a lot of code in there is duplicated from other modules so it might make sense that they share the code instead.
Finally I wonder if the PR update code could be simplified by simply fetching the PR branch, updating the opam files there to properly set `url.src`. This would probably make the code simpler and also improve the feature as it would benefit from local changes to the PR, e.g. when the user updated the depends field to get a green CI. If that's too much work we can do it in a separate PR, after releasing the draft mode as it is!

gpetiot commented on bin/undraft.ml:
```diff
@@ -0,0 +1,222 @@
+  let url = OpamUrl.parse url in
+  Pkg.distrib_file ~dry_run pkg >>= fun distrib_file ->
+  let file = Fpath.to_string distrib_file in
+  let hash algo = OpamHash.compute ~kind:algo file in
```
It is not necessary indeed, done!

gpetiot commented on bin/undraft.ml:
```diff
@@ -0,0 +1,222 @@
+    (* git add *)
+    Vcs.run_git_quiet vcs ~dry_run ~force:true Cmd.(v "add" % p dst)
+  in
+  let rec prepare_packages = function
```
Indeed, done!

gpetiot commented on bin/undraft.ml:
```diff
@@ -0,0 +1,222 @@
+    let dir = name ^ "." ^ version in
+    let src = Fpath.(cwd / "_build" / dir) in
+    let dst = Fpath.(v "packages" / name / dir) in
+    let cp f =
```
I just realized there was already a Sos.cp function, done!

gpetiot commented on bin/undraft.ml:
```diff
@@ -0,0 +1,222 @@
+  let branch = Fmt.strf "release-%s-%s" pkg_name tag in
+  Vcs.get () >>= fun vcs ->
+  OS.Dir.current () >>= fun cwd ->
+  let prepare_package name =
```
done!

gpetiot commented on bin/undraft.ml:
```diff
@@ -0,0 +1,222 @@
+      | None -> R.error_msg "Unknown local repository."))
+  >>= fun local_repo ->
+  (match remote_repo with
+  | Some r -> Ok r
```
I tried a few things but the mixing of the result and the option types make it not easier to read. I think it can be improved in a following PR, I would like to do something to what we have in ocamlformat, that is: a Config.t type that is built by overriding the default config with environment, .ocamlformat files and options passed by the user, for dnue-release it can be: current config.t read from the yaml file + command line options, it should be computed once at the beginning of each command and then just passed down each function wihtout having to worry if we have taken each option into account.
What do you think?

(The other way would be to dramatically reduce the command line options of dune-release)

gpetiot commented on bin/opam.ml:
```diff
@@ -178,6 +182,17 @@ let submit ?distrib_uri ~token ~dry_run ~yes ~opam_repo ~user local_repo
+  Pkg.name pkg >>= fun name ->
+  (if draft then Ok ()
+  else
+    match Config.Draft_release.is_set ~dry_run ~build_dir ~name ~version with
```
I agree, it can be done in a future PR once users start asking for this feature.

NathanReb reviewed (approved):
Let's merge this!
